Notes on running on DGX Spark [WIP]

We just got some DGX Spark machines at DMI. I am writing down some notes here on what I have done to get it working.

I've put any changes I have made to the codebase (I'm starting from `main`) on https://github.com/leifdenby/neural-lam/tree/dev/dgx-spark

Steps to successfully run tests on DGX Spark GPU:

- there is currently no wheel for `cartopy` for `python3.12`, so I created venv with `python3.11` with `uv sync -p 3.11`
- installed `dev` packages (`uv sync --extra dev`) and ran tests with `uv run python -m pytest`, ran successfully, but didn't pick up GPU
- installed pytorch `2.9.0+cu130` with `uv pip install --index-url https://download.pytorch.org/whl/cu130 torch`
   - torch now picks up `GB10` GPU, but I get warning that 
```
    Found GPU0 NVIDIA GB10 which is of cuda capability 12.1.
    Minimum and Maximum cuda capability supported by this version of PyTorch is
    (8.0) - (12.0)
```
  - to run training tests, had to switch to use a single GPU (`devices=1` rather than `devices=2` in `tests/test_training.py`) and comment out `self.aggregate_and_plot_metrics(self.val_metrics, prefix="val")` because we still have a bug with single device training: https://github.com/mllam/neural-lam/blob/579efc8050d0f51c04c6b369f34126ac29e41106/tests/test_training.py#L45
  - training tests do seem to complete rather slowly though, may have to use docker container as suggested by nvidia

Trying inside docker container:
- add myself to `docker` group with `sudo usermod -a -G docker lcd`
- start interactive shell in docker container using image from nvidia:
```bash
docker run --gpus all -it --rm --ipc=host \
-v ${PWD}:/workspace -w /workspace \
nvcr.io/nvidia/pytorch:25.09-py3
```
- install all deps with pip `pip install .`
- run tests `python -m pytest -s`, this brought up a problem with test `test_training[mdp]`:
```

FAILED tests/test_training.py::test_training[mdp] - IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)
```

```bash
    @pytest.mark.parametrize("datastore_name", DATASTORES.keys())
    def test_training(datastore_name):
        datastore = init_datastore_example(datastore_name)

        if not isinstance(datastore, BaseRegularGridDatastore):
            pytest.skip(
                f"Skipping test for {datastore_name} as it is not a regular "
                "grid datastore."
            )

>       run_simple_training(datastore, set_output_std=False)

tests/test_training.py:121:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_training.py:108: in run_simple_training
    trainer.fit(model=model, datamodule=data_module)
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/trainer/trainer.py:560: in fit
    call._call_and_handle_interrupt(
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/trainer/call.py:49: in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/trainer/trainer.py:598: in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/trainer/trainer.py:1011: in _run
    results = self._run_stage()
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/trainer/trainer.py:1053: in _run_stage
    self._run_sanity_check()
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/trainer/trainer.py:1082: in _run_sanity_check
    val_loop.run()
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/loops/utilities.py:179: in _decorator
    return loop_run(self, *args, **kwargs)
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/loops/evaluation_loop.py:145: in run
    self._evaluation_step(batch, batch_idx, dataloader_idx, dataloader_iter)
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/loops/evaluation_loop.py:437: in _evaluation_step
    output = call._call_strategy_hook(trainer, hook_name, *step_args)
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/trainer/call.py:329: in _call_strategy_hook
    output = fn(*args, **kwargs)
/usr/local/lib/python3.12/dist-packages/pytorch_lightning/strategies/strategy.py:412: in validation_step
    return self.lightning_module.validation_step(*args, **kwargs)
neural_lam/models/ar_model.py:331: in validation_step
    prediction, target, pred_std, _ = self.common_step(batch)
neural_lam/models/ar_model.py:282: in common_step
    prediction, pred_std = self.unroll_prediction(
neural_lam/models/ar_model.py:240: in unroll_prediction
    pred_state, pred_std = self.predict_step(
neural_lam/models/base_graph_model.py:325: in predict_step
    mesh_rep = self.g2m_gnn(
/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py:1775: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py:1786: in _call_impl
    return forward_call(*args, **kwargs)
neural_lam/interaction_net.py:103: in forward
    edge_rep_aggr, edge_diff = self.propagate(
/usr/local/lib/python3.12/dist-packages/torch_geometric/nn/conv/message_passing.py:484: in propagate
    out = self.aggregate(out, **aggr_kwargs)
neural_lam/interaction_net.py:130: in aggregate
    aggr = super().aggregate(inputs, index, ptr, self.num_rec)
/usr/local/lib/python3.12/dist-packages/torch_geometric/nn/conv/message_passing.py:608: in aggregate
    return self.aggr_module(inputs, index, ptr=ptr, dim_size=dim_size,
/usr/local/lib/python3.12/dist-packages/torch_geometric/nn/aggr/base.py:116: in __call__
    raise e
/usr/local/lib/python3.12/dist-packages/torch_geometric/nn/aggr/base.py:109: in __call__
    return super().__call__(x, index, ptr, dim_size, dim, **kwargs)
/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py:1775: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py:1786: in _call_impl
    return forward_call(*args, **kwargs)
/usr/local/lib/python3.12/dist-packages/torch_geometric/nn/aggr/basic.py:21: in forward
    return self.reduce(x, index, ptr, dim_size, dim, reduce='sum')
/usr/local/lib/python3.12/dist-packages/torch_geometric/nn/aggr/base.py:155: in reduce
    return scatter(x, index, dim, dim_size, reduce)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

src = tensor([[[-1.0756, -0.9201,  0.9494,  1.0463],
         [-1.0760, -0.9190,  0.9374,  1.0576],
         [-1.0872, -0.90...9900],
         [-1.1252, -0.8662,  1.0032,  0.9882],
         [-1.1081, -0.8845,  0.9508,  1.0417]]], device='cuda:0')
index = tensor([[[  0,   0,   0,   0],
         [  0,   0,   0,   0],
         [  0,   0,   0,   0],
         ...,
         [7...   ...,
         [728, 728, 728, 728],
         [728, 728, 728, 728],
         [728, 728, 728, 728]]], device='cuda:0')
dim = 1, dim_size = tensor(729), reduce = 'sum'

    def scatter(src: Tensor, index: Tensor, dim: int = 0,
                dim_size: Optional[int] = None, reduce: str = 'sum') -> Tensor:
        r"""Reduces all values from the :obj:`src` tensor at the indices
        specified in the :obj:`index` tensor along a given dimension
        :obj:`dim`. See the `documentation
        <https://pytorch-scatter.readthedocs.io/en/latest/functions/
        scatter.html>`__ of the :obj:`torch_scatter` package for more
        information.

        Args:
            src (torch.Tensor): The source tensor.
            index (torch.Tensor): The index tensor.
            dim (int, optional): The dimension along which to index.
                (default: :obj:`0`)
            dim_size (int, optional): The size of the output tensor at
                dimension :obj:`dim`. If set to :obj:`None`, will create a
                minimal-sized output tensor according to
                :obj:`index.max() + 1`. (default: :obj:`None`)
            reduce (str, optional): The reduce operation (:obj:`"sum"`,
                :obj:`"mean"`, :obj:`"mul"`, :obj:`"min"` or :obj:`"max"`).
                (default: :obj:`"sum"`)
        """
        if index.dim() != 1:
            raise ValueError(f"The `index` argument must be one-dimensional "
                             f"(got {index.dim()} dimensions)")

        dim = src.dim() + dim if dim < 0 else dim

        if dim < 0 or dim >= src.dim():
            raise ValueError(f"The `dim` argument must lay between 0 and "
                             f"{src.dim() - 1} (got {dim})")

        if dim_size is None:
            dim_size = int(index.max()) + 1 if index.numel() > 0 else 0

        # For now, we maintain various different code paths, based on whether
        # the input requires gradients and whether it lays on the CPU/GPU.
        # For example, `torch_scatter` is usually faster than
        # `torch.scatter_reduce` on GPU, while `torch.scatter_reduce` is faster
        # on CPU.
        # `torch.scatter_reduce` has a faster forward implementation for
        # "min"/"max" reductions since it does not compute additional arg
        # indices, but is therefore way slower in its backward implementation.
        # More insights can be found in `test/utils/test_scatter.py`.

        size = list(src.size())
        size[dim] = dim_size

        # For "sum" and "mean" reduction, we make use of `scatter_add_`:
        if reduce == 'sum' or reduce == 'add':
            index = broadcast(index, src, dim)
>           return src.new_zeros(size).scatter_add_(dim, index, src)
E           IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -72,6 +72,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Maintenance
 
+- Add a short README pointer to [\#163](https://github.com/mllam/neural-lam/issues/163) for DGX Spark / PyTorch container compatibility notes, so users hitting `torch_scatter` errors know where to find the known-working / known-failing combos [\#266](https://github.com/mllam/neural-lam/pull/266) @Jayant-kernel
+
 - Group the existing Neural-LAM citation papers in the README under a `### Core Neural-LAM Publications` subheading for clearer structure [\#633](https://github.com/mllam/neural-lam/pull/633) @HetaviM29
 
 - Add unit tests for `inverse_softplus` covering roundtrip identity (parametrized over `beta`), near-zero clamping, and above-threshold linear passthrough [\#419](https://github.com/mllam/neural-lam/pull/419) @Riteesh-NITT

--- a/README.md
+++ b/README.md
@@ -504,6 +504,8 @@ srun -ul python -m neural_lam.train_model \
 
 When using on a system without SLURM, where all GPU's are visible, it is possible to select a subset of GPU's to use for training with the `devices` cli argument, e.g. `--devices 0 1` to use the first 2 GPU's.
 
+> **DGX Spark / container compatibility:** see [issue #163](https://github.com/mllam/neural-lam/issues/163) for a list of currently known-working and known-failing PyTorch / CUDA / container combinations on DGX Spark.
+
 ## Evaluate Models
 Evaluation is also done using `python -m neural_lam.train_model --config_path <config-path>`, but using the `--eval` option.
 Use `--eval val` to evaluate the model on the validation set and `--eval test` to evaluate on test data.

